### PR TITLE
[WIP] split read and write SOT types

### LINKED ISSUE #559: [BUG] Fix breaking changes with source of truth
Store 4 had a source of truth that can write one type and read another. During some work adding mutable store, we broke that contract. I'll be adding the functionality back, it was an unintentional oversight

### PR DESCRIPTION
**Pull request by digitalbuddha**
Need to fix some of the newer tests and document. Fixes #559  #558

### CODE REVIEW

File: rx2/src/main/kotlin/org/mobilenativefoundation/store/rx2/RxSourceOfTruth.kt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package org.mobilenativefoundation.store.rx2
 
 
+import android.icu.util.Output
```
Comment (matt-ramotar) [inline rx2/src/main/kotlin/org/mobilenativefoundation/store/rx2/RxSourceOfTruth.kt:4]: Is this intentional? It looks like it's not being used elsewhere

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt [hunk 2]
```diff
@@ -1,47 +1,45 @@
...
-        private var fromNetworkToOutput: ((network: Network) -> Output)? = null
-        private var fromLocalToOutput: ((local: Local) -> Output)? = null
-
-        fun build(): Converter<Network, Output, Local> =
-            RealConverter(fromOutputToLocal, fromNetworkToOutput, fromLocalToOutput)
-
-        fun fromOutputToLocal(converter: (output: Output) -> Local): Builder<Network, Output, Local> {
+/**
+ * Converter is a utility interface that can be used to convert a network or output model to a local model.
+ * Network to Model conversion is needed when the network model is different what you are saving in
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt:5]: > Network to Model

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt [hunk 3]
```diff
@@ -1,47 +1,45 @@
...
+ * your Source of Truth.
+ * Output to Local conversion is needed when you are doing local writes in a MutableStore
+ * @param Network the network model type
+ * @param Output the output model type
+ * @param Local the local model type
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt:10]: I think these definitions should be expanded. For example
```
@param Network The network data source model type. This is the type used in [Fetcher]
@param Output The common model type emitted from Store
@param Local The local data source model type. This is the type used by [SourceOfTruth] and [MemoryCache]
```

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt [hunk 4]
```diff
@@ -1,47 +1,45 @@
...
+ */
+interface Converter<Network : Any, Local : Any, Output : Any> {
+    fun fromNetworkToLocal(network: Network): Local
+    fun fromOutputToLocal(output: Output): Local
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/Converter.kt:14]: My understanding is `OutputToLocal` is only used with `MutableStore`. Should this interface be separated?

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt [hunk 5]
```diff
@@ -55,7 +56,16 @@ internal class FetcherController<Key : Any, Network : Any, Output : Any, Local :
      */
     private val sourceOfTruth: SourceOfTruthWithBarrier<Key, Network, Output, Local>?,
 
-    private val converter: Converter<Network, Output, Local>? = null
+    private val converter: Converter<Network, Local, Output> = object : Converter<Network, Local, Output> {
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt:59]: Can you add a comment explaining why it is safe to use this object as the converter?

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/RealStoreBuilder.kt [hunk 6]
```diff
@@ -16,30 +16,39 @@ import org.mobilenativefoundation.store.store5.StoreBuilder
...
 
 internal class RealStoreBuilder<Key : Any, Network : Any, Output : Any, Local : Any>(
     private val fetcher: Fetcher<Key, Network>,
-    private val sourceOfTruth: SourceOfTruth<Key, Local>? = null,
-    private val memoryCache: Cache<Key, Output>? = null
+    private val sourceOfTruth: SourceOfTruth<Key, Local, Output>? = null,
+    private val memoryCache: Cache<Key, Output>? = null,
+    private val converter: Converter<Network, Local, Output> = object :
+        Converter<Network, Local, Output> {
+        override fun fromOutputToLocal(output: Output): Local =
```
Comment (matt-ramotar) [inline store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/RealStoreBuilder.kt:44]: This `fromOutputToLocal` method is only invoked internally right? Should the error message convey it's not an issue with their usage?

**Review by matt-ramotar — APPROVED** [review]
Overall LGTM. I wonder if we should separate the converter for Store and MutableStore usage. I think when we support #554 users would otherwise find `fromOutputToLocal` confusing. The failing test is with generics in rx2

File: store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt (hunk 5 above)
Comment (digitalbuddha) [reply store/src/commonMain/kotlin/org/mobilenativefoundation/store/store5/impl/FetcherController.kt:59]: This is due to a bit of tangling between store and mutable store.  In store, there is no concept of local, the network type is always the same as what is going into the source of truth. This default parameter keeps us from having to pass in that dummy everywhere. The second function `fromOutputToLocal` will never be called, I'll throw an exception from there.

**Comment by digitalbuddha** [comment]
@matt-ramotar UpdaterTests was failing for me locally, I don't think it was rx related

**Pull request merged**